ethereum/core/src/main/java/org/hyperledger/besu/ethereum/mainnet/UnifiedProtocolSchedule.java

```diff
@@ -31,64 +34,81 @@
 
 import com.google.common.annotations.VisibleForTesting;
 
-public class DefaultTimestampSchedule implements TimestampSchedule {
+public class UnifiedProtocolSchedule implements ProtocolSchedule {
 
   @VisibleForTesting
   protected NavigableSet<ScheduledProtocolSpec> protocolSpecs =
       new TreeSet<>(Comparator.comparing(ScheduledProtocolSpec::milestone).reversed());
 
   private final Optional<BigInteger> chainId;
 
-  DefaultTimestampSchedule(final Optional<BigInteger> chainId) {
+  public UnifiedProtocolSchedule(final Optional<BigInteger> chainId) {
     this.chainId = chainId;
   }
 
   @VisibleForTesting
-  protected DefaultTimestampSchedule(final DefaultTimestampSchedule timestampSchedule) {
-    this.chainId = timestampSchedule.chainId;
-    this.protocolSpecs = timestampSchedule.protocolSpecs;
+  protected UnifiedProtocolSchedule(final UnifiedProtocolSchedule protocolSchedule) {
+    this.chainId = protocolSchedule.chainId;
+    this.protocolSpecs = protocolSchedule.protocolSpecs;
   }
 
   @Override
-  public Optional<ProtocolSpec> getByTimestamp(final long timestamp) {
-    for (final ScheduledProtocolSpec protocolSpec : protocolSpecs) {
-      if (protocolSpec.milestone() <= timestamp) {
-        return Optional.of(protocolSpec.spec());
+  public ProtocolSpec getByBlockHeader(final ProcessableBlockHeader blockHeader) {
+    // TODO SLD to remove once it's been discussed in PR review
+    //    checkArgument(number >= 0, "number must be non-negative");
+    checkArgument(
+        !protocolSpecs.isEmpty(), "At least 1 milestone must be provided to the protocol schedule");
+    checkArgument(
+        protocolSpecs.last().milestone() == 0, "There must be a milestone starting from block 0");
+
+    // protocolSpecs is sorted in descending block order, so the first one we find that's lower than
+    // the requested level will be the most appropriate spec
+    for (final ScheduledProtocolSpec scheduledProtocolSpec : protocolSpecs) {
+      if (scheduledProtocolSpec.isTimestampMilestone()) {
+        if (blockHeader.getTimestamp() >= scheduledProtocolSpec.milestone()) {
+          return scheduledProtocolSpec.spec();
+        }
+      } else {
+        if (blockHeader.getNumber() >= scheduledProtocolSpec.milestone()) {
+          return scheduledProtocolSpec.spec();
+        }
```

Comment:
This is the most important change and represents a change to the algorithm for retrieving the protocol schedule: it still uses the original MutableScheduleProtocol.getByBlockNumber algorithm (running through the schedule in reverse order and returning as soon as our input milestone is beyond a spec), however it checks the timestamp first, followed by the blockNumber.

Reply:
This is the original MutableProtocolSchedule.getByBlockNumber algorithm for comparison:

```
  @Override
  public ProtocolSpec getByBlockNumber(final long number) {
    checkArgument(number >= 0, "number must be non-negative");
    checkArgument(
        !protocolSpecs.isEmpty(), "At least 1 milestone must be provided to the protocol schedule");
    checkArgument(
        protocolSpecs.last().milestone() == 0, "There must be a milestone starting from block 0");
    // protocolSpecs is sorted in descending block order, so the first one we find that's lower than
    // the requested level will be the most appropriate spec
    for (final ScheduledProtocolSpec s : protocolSpecs) {
      if (number >= s.milestone()) {
        return s.spec();
      }
    }
    return null;
  }
  ```

Reply:
nit: you could make this closer to the original and remove a little bit of dupe by doing something like


```
    for (final ScheduledProtocolSpec scheduledProtocolSpec : protocolSpecs) {
      final long milestone =
          scheduledProtocolSpec.isTimestampMilestone()
              ? blockHeader.getTimestamp()
              : blockHeader.getNumber();
      if (milestone >= scheduledProtocolSpec.milestone()) {
        return scheduledProtocolSpec.spec();
      }
    }
```